pkg/jobs/hive/hive.go

```diff
@@ -329,8 +333,268 @@ func UpgradeCluster(client clientv1.Client, clusterName string, curator *cluster
+	// before upgrading or else upgrade will be blocked.
+	ocpConfigMCV := &managedclusterviewv1beta1.ManagedClusterView{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterName + "admack",
+			Namespace: clusterName,
+			Labels: map[string]string{
+				MCVUpgradeLabel: clusterName,
+			},
+		},
+		Spec: managedclusterviewv1beta1.ViewSpec{
+			Scope: managedclusterviewv1beta1.ViewScope{
+				Kind:      "ConfigMap",
+				Name:      "admin-acks",
+				Namespace: "openshift-config",
+				Version:   "v1",
+			},
+		},
+	}
+
+	ocpConfigView := managedclusterviewv1beta1.ManagedClusterView{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Name:      clusterName + "admack",
+		Namespace: clusterName,
+	}, &ocpConfigView); err != nil {
+		// check if mcv exists before creating
+
+		klog.V(2).Info("Create managedclusterview " + clusterName + "admack")
+		if err := client.Create(context.TODO(), ocpConfigMCV); err != nil {
+			return err
+		}
+	}
+
+	ocpConfigGetErr := errors.New("Failed to get remote admin-ack configmap")
+	resultOCPConfigMCV := managedclusterviewv1beta1.ManagedClusterView{}
+	if err := waitForMCV(client, clusterName+"admack", clusterName, &resultOCPConfigMCV, ocpConfigGetErr); err != nil {
+		return err
+	}
+
+	resultConfigMap := resultOCPConfigMCV.Status.Result
+	configMap := map[string]interface{}{}
+
+	if resultConfigMap.Raw != nil {
+		err := json.Unmarshal(resultConfigMap.Raw, &configMap)
+		utils.CheckError(err)
+	} else {
+		return ocpConfigGetErr
+	}
+
+	// Get clusterversion from managed cluster to initiate upgrade
+	managedclusterview := &managedclusterviewv1beta1.ManagedClusterView{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: clusterName,
+			Labels: map[string]string{
+				MCVUpgradeLabel: clusterName,
+			},
+		},
+		Spec: managedclusterviewv1beta1.ViewSpec{
+			Scope: managedclusterviewv1beta1.ViewScope{
+				Group:     "config.openshift.io",
+				Kind:      "ClusterVersion",
+				Name:      "version",
+				Namespace: "",
+				Version:   "v1",
+			},
+		},
+	}
+
+	mcview := managedclusterviewv1beta1.ManagedClusterView{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: clusterName,
+		Name:      clusterName,
+	}, &mcview); err != nil {
+
+		klog.V(2).Info("Create managedclusterview " + clusterName)
+		if err := client.Create(context.TODO(), managedclusterview); err != nil {
+			return err
+		}
+	}
+
+	resultmcview := managedclusterviewv1beta1.ManagedClusterView{}
+	if err := waitForMCV(client, clusterName, clusterName, &resultmcview, getErr); err != nil {
+		return err
+	}
+
+	resultClusterVersion := resultmcview.Status.Result
+	clusterVersion := map[string]interface{}{}
+
+	if resultClusterVersion.Raw != nil {
+		err := json.Unmarshal(resultClusterVersion.Raw, &clusterVersion)
+		utils.CheckError(err)
+	} else {
+		return getErr
+	}
+
+	currentOCPVer := clusterVersion["status"].(map[string]interface{})["desired"].(map[string]interface{})["version"].(string)
```

Comment:
yikes :sweat_smile: 

is this copy from somewhere? lots of nilpointer potentials here.

Reply:
Yeah it's used a lot in this area 😅
https://github.com/stolostron/cluster-curator-controller/blob/main/pkg/jobs/hive/hive.go#L370